File: typescript/packages/mechanisms/hypercore/src/exact/facilitator/scheme.ts
import type {
  SchemeNetworkFacilitator,
  PaymentPayload,
  PaymentRequirements,
  VerifyResponse,
  SettleResponse,
  Network,
} from "@x402/core/types";
import type {
  HypercorePaymentPayload,
  HyperliquidApiResponse,
  HypercoreSendAssetAction,
} from "../../types.js";
import { recoverTypedDataAddress } from "viem";
import {
  HYPERCORE_EIP712_DOMAIN,
  HYPERCORE_EIP712_TYPES,
  MAX_NONCE_AGE_MS,
  TX_HASH_LOOKUP,
  HYPERCORE_NETWORK_CONFIGS,
  HYPERCORE_API_URLS_BY_NETWORK,
} from "../../constants.js";

/**
 * Hypercore scheme facilitator implementation.
 */
export class ExactHypercoreScheme implements SchemeNetworkFacilitator {
  readonly scheme = "exact";
  readonly caipFamily = "hypercore:*";

  /**
   * Create a facilitator scheme.
   *
   * @param config - Optional configuration object.
   * @param config.apiUrl - Optional API URL override for the network.
   */
  constructor(private config: { apiUrl?: string } = {}) {}

  /**
   * Return extra metadata for the network.
   *
   * @param _ - Network identifier.
   * @returns Extra metadata, if any.
   */
  getExtra(_: Network): Record<string, unknown> | undefined {
    return undefined;
  }

  /**
   * Return signer addresses for the network.
   *
   * @param _ - Network identifier.
   * @returns Signer addresses.
   */
  getSigners(_: string): string[] {
    return [];
  }

  /**
   * Verify a Hypercore payment payload against requirements.
   *
   * @param payload - The payment payload to verify.
   * @param requirements - The payment requirements to verify against.
   * @returns Verification result.
   */
  async verify(
    payload: PaymentPayload,
    requirements: PaymentRequirements,
  ): Promise<VerifyResponse> {
    const hypercorePayload = payload.payload as unknown as HypercorePaymentPayload;

    if (!requirements.network.startsWith("hypercore:")) {
      return {
        isValid: false,
        invalidReason: `Invalid network: ${requirements.network}. Expected hypercore:mainnet or hypercore:testnet`,
      };
    }

    const config = HYPERCORE_NETWORK_CONFIGS[requirements.network];
    if (!config) {
      return {
        isValid: false,
        invalidReason: `Invalid network: ${requirements.network}. Expected hypercore:mainnet or hypercore:testnet`,
      };
    }

    if (hypercorePayload.action.type !== "sendAsset") {
      return {
        isValid: false,
        invalidReason: `Invalid action type: ${hypercorePayload.action.type}. Expected sendAsset`,
      };
    }

    if (hypercorePayload.action.destination.toLowerCase() !== requirements.payTo.toLowerCase()) {
      return {
        isValid: false,
        invalidReason: `Destination mismatch: ${hypercorePayload.action.destination} vs ${requirements.payTo}`,
      };
    }

    const decimals = config.defaultAsset.decimals;
    const payloadAmount = this.parseAmountToInteger(hypercorePayload.action.amount, decimals);
    const requiredAmount = parseInt(requirements.amount);

    if (payloadAmount < requiredAmount) {
      return {
        isValid: false,
        invalidReason: `Insufficient amount: ${payloadAmount} < ${requiredAmount}`,
      };
    }

    if (requirements.asset && hypercorePayload.action.token !== requirements.asset) {
      return {
        isValid: false,
        invalidReason: `Token mismatch: ${hypercorePayload.action.token} vs ${requirements.asset}`,
      };
    }

    const now = Date.now();
    const nonceAge = now - hypercorePayload.nonce;

    if (nonceAge > MAX_NONCE_AGE_MS) {
      return {
        isValid: false,
        invalidReason: `Nonce too old: ${nonceAge}ms (max ${MAX_NONCE_AGE_MS}ms)`,
      };
    }

    if (
      !hypercorePayload.signature ||
      !hypercorePayload.signature.r ||
      !hypercorePayload.signature.s ||
      typeof hypercorePayload.signature.v !== "number"
    ) {
      return {
        isValid: false,
        invalidReason: "Invalid signature structure",
      };
    }

    return { isValid: true };
  }

  /**
   * Settle a Hypercore payment using the Hyperliquid API.
   *
   * @param payload - The payment payload to settle.
   * @param requirements - The payment requirements.
   * @returns Settlement result.
   */
  async settle(
    payload: PaymentPayload,
    requirements: PaymentRequirements,
  ): Promise<SettleResponse> {
    const verifyResult = await this.verify(payload, requirements);
    if (!verifyResult.isValid) {
      throw new Error(`Verification failed: ${verifyResult.invalidReason}`);
    }

    const hypercorePayload = payload.payload as unknown as HypercorePaymentPayload;
    const apiUrl = this.getApiUrl(requirements.network);

    try {
      const payer = await this.recoverPayer(hypercorePayload.action, hypercorePayload.signature);

      const startTime = Date.now();

      const response = await fetch(`${apiUrl}/exchange`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          action: hypercorePayload.action,
          nonce: hypercorePayload.nonce,
          signature: hypercorePayload.signature,
          vaultAddress: null,
        }),
      });

      if (!response.ok) {
        throw new Error(`Hyperliquid API error: ${response.status} ${response.statusText}`);
      }

      const result = (await response.json()) as HyperliquidApiResponse;

      if (result.status !== "ok") {
        throw new Error(`Settlement failed: ${JSON.stringify(result)}`);
      }

      const transactionHash = await this.getTransactionHash(
        apiUrl,
        payer,
        hypercorePayload.action.destination,
        hypercorePayload.nonce,
        startTime,
      );

      return {
        success: true,
        transaction: transactionHash,
        network: requirements.network,
        payer,
      };
    } catch (error) {
      throw new Error(
        `Hypercore settlement failed: ${error instanceof Error ? error.message : "Unknown error"}`,
      );
    }
  }

  /**
   * Get the API URL for a specific network.
   *
   * @param network - Network identifier (e.g. hypercore:mainnet, hypercore:testnet).
   * @returns API URL for the network.
   */
  private getApiUrl(network: string): string {
    const url = HYPERCORE_API_URLS_BY_NETWORK[network] ?? this.config.apiUrl;
    if (!url) {
      throw new Error(`No API URL configured for network: ${network}`);
    }
    return url;
  }

  /**
   * Recover the payer address from a typed-data signature.
   *
   * @param action - The SendAsset action that was signed.
   * @param signature - Signature parts.
   * @param signature.r - Signature r value.
   * @param signature.s - Signature s value.
   * @param signature.v - Signature v value.
   * @returns Payer address.
   */
  private async recoverPayer(
    action: HypercoreSendAssetAction,
    signature: { r: string; s: string; v: number },
  ): Promise<string> {
    const address = await recoverTypedDataAddress({
      domain: {
        ...HYPERCORE_EIP712_DOMAIN,
        chainId: Number(HYPERCORE_EIP712_DOMAIN.chainId),
      },
      types: HYPERCORE_EIP712_TYPES,
      primaryType: "HyperliquidTransaction:SendAsset",
      message: {
        ...action,
        nonce: BigInt(action.nonce),
      },
      signature:
        `${signature.r}${signature.s.slice(2)}${signature.v.toString(16).padStart(2, "0")}` as `0x${string}`,
    });
    return address;
  }

  /**
   * Fetch the transaction hash from ledger updates.
   *
   * @param apiUrl - Hyperliquid API URL to query.
   * @param user - The user's address who sent the transaction.
   * @param destination - Destination address.
   * @param nonce - Payment nonce.
   * @param startTime - Submission timestamp (ms).
   * @returns Transaction hash.
   */
  private async getTransactionHash(
    apiUrl: string,
    user: string,
    destination: string,
    nonce: number,
    startTime: number,
  ): Promise<string> {
    for (let attempt = 0; attempt < TX_HASH_LOOKUP.maxRetries; attempt++) {
      try {
        if (attempt > 0) {
          await new Promise(resolve => setTimeout(resolve, TX_HASH_LOOKUP.retryDelay));
        }

        const response = await fetch(`${apiUrl}/info`, {
          method: "POST",
          headers: {
            "Content-Type": "application/json",
          },
          body: JSON.stringify({
            type: "userNonFundingLedgerUpdates",
            user: user,
            startTime: startTime - TX_HASH_LOOKUP.lookbackWindow,
          }),
        });

        if (!response.ok) {
          throw new Error(`Failed to query ledger updates: ${response.status}`);
        }

        const ledgerUpdates = (await response.json()) as Array<{
          time: number;
          hash: string;
          delta: {
            [key: string]: unknown;
            type: string;
            destination?: string;
            nonce?: number;
          };
        }>;

        console.log(
          `[Hypercore] Attempt ${attempt + 1}: Found ${ledgerUpdates.length} ledger updates for user ${user}`,
        );

        const transfer = ledgerUpdates.find(
          update =>
            update.delta.type === "send" &&
            update.delta.destination?.toLowerCase() === destination.toLowerCase() &&
            update.delta.nonce === nonce,
        );

        if (transfer) {
          console.log(`[Hypercore] Found transaction hash: ${transfer.hash}`);
          return transfer.hash;
        }

        if (attempt < TX_HASH_LOOKUP.maxRetries - 1) {
          console.log(
            `[Hypercore] Transaction not indexed yet, retrying in ${TX_HASH_LOOKUP.retryDelay}ms...`,
          );
        }
      } catch (error) {
        console.error(`[Hypercore] Error querying ledger updates (attempt ${attempt + 1}):`, error);
        if (attempt === TX_HASH_LOOKUP.maxRetries - 1) {
          throw error;
        }
      }
    }

    throw new Error(
      `Transaction hash not found in ledger updates after ${TX_HASH_LOOKUP.maxRetries} attempts`,
    );
  }

  /**
   * Parse a USD amount string into a 6-decimal integer.
   *
   * @param amount - USD amount string.
   * @param decimals - Number of decimals for the token.
   * @returns Amount as integer.
   */
  private parseAmountToInteger(amount: string, decimals: number): number {
    const num = parseFloat(amount);
    return Math.floor(num * Math.pow(10, decimals));
  }
}

File: go/mechanisms/hypercore/exact/facilitator/scheme.go
package facilitator

import (
	"bytes"
	"context"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"math/big"
	"net/http"
	"strings"
	"time"

	"github.com/ethereum/go-ethereum/crypto"

	x402 "github.com/coinbase/x402/go"
	"github.com/coinbase/x402/go/mechanisms/evm"
	"github.com/coinbase/x402/go/mechanisms/hypercore"
	"github.com/coinbase/x402/go/types"
)

type ExactHypercoreScheme struct {
	apiURL string // optional override
}

// NewExactHypercoreScheme creates a new scheme. Pass "" for apiURL to use built-in defaults.
func NewExactHypercoreScheme(apiURL ...string) *ExactHypercoreScheme {
	s := &ExactHypercoreScheme{}
	if len(apiURL) > 0 {
		s.apiURL = apiURL[0]
	}
	return s
}

func (f *ExactHypercoreScheme) Scheme() string {
	return hypercore.SchemeExact
}

func (f *ExactHypercoreScheme) CaipFamily() string {
	return "hypercore:*"
}

// getAPIURL returns the API URL for a specific network, falling back to the configured default.
func (f *ExactHypercoreScheme) getAPIURL(network string) string {
	if url, ok := hypercore.NetworkAPIURLs[network]; ok {
		return url
	}
	return f.apiURL
}

func (f *ExactHypercoreScheme) GetExtra(network x402.Network) map[string]interface{} {
	return nil
}

func (f *ExactHypercoreScheme) GetSigners(network x402.Network) []string {
	return []string{}
}

func (f *ExactHypercoreScheme) Verify(
	ctx context.Context,
	payload types.PaymentPayload,
	requirements types.PaymentRequirements,
) (*x402.VerifyResponse, error) {
	hypercorePayload, err := parsePayload(payload.Payload)
	if err != nil {
		return &x402.VerifyResponse{
			IsValid:       false,
			InvalidReason: ErrInvalidPayloadStructure,
		}, err
	}

	if !strings.HasPrefix(string(requirements.Network), "hypercore:") {
		return &x402.VerifyResponse{
			IsValid:       false,
			InvalidReason: fmt.Sprintf("%s: %s", ErrInvalidNetwork, requirements.Network),
		}, nil
	}

	config, ok := hypercore.NetworkConfigs[string(requirements.Network)]
	if !ok {
		return &x402.VerifyResponse{
			IsValid:       false,
			InvalidReason: fmt.Sprintf("%s: %s", ErrInvalidNetwork, requirements.Network),
		}, nil
	}

	if hypercorePayload.Action.Type != "sendAsset" {
		return &x402.VerifyResponse{
			IsValid:       false,
			InvalidReason: fmt.Sprintf("%s: %s", ErrInvalidActionType, hypercorePayload.Action.Type),
		}, nil
	}

	if !strings.EqualFold(hypercorePayload.Action.Destination, requirements.PayTo) {
		return &x402.VerifyResponse{
			IsValid:       false,
			InvalidReason: ErrDestinationMismatch,
		}, nil
	}

	payloadAmount, err := hypercore.ParseAmountToInteger(hypercorePayload.Action.Amount, config.DefaultAsset.Decimals)
	if err != nil {
		return &x402.VerifyResponse{
			IsValid:       false,
			InvalidReason: ErrInvalidAmountFormat,
		}, err
	}

	requiredAmount := new(big.Int)
	requiredAmount.SetString(requirements.Amount, 10)

	if payloadAmount.Cmp(requiredAmount) < 0 {
		return &x402.VerifyResponse{
			IsValid:       false,
			InvalidReason: ErrInsufficientAmount,
		}, nil
	}

	if requirements.Asset != "" && hypercorePayload.Action.Token != requirements.Asset {
		return &x402.VerifyResponse{
			IsValid:       false,
			InvalidReason: ErrTokenMismatch,
		}, nil
	}

	if !hypercore.IsNonceFresh(hypercorePayload.Nonce, time.Duration(hypercore.MaxNonceAgeSeconds)*time.Second) {
		return &x402.VerifyResponse{
			IsValid:       false,
			InvalidReason: ErrNonceTooOld,
		}, nil
	}

	if hypercorePayload.Signature.R == "" || hypercorePayload.Signature.S == "" {
		return &x402.VerifyResponse{
			IsValid:       false,
			InvalidReason: ErrInvalidSignature,
		}, nil
	}

	return &x402.VerifyResponse{
		IsValid: true,
	}, nil
}

func (f *ExactHypercoreScheme) Settle(
	ctx context.Context,
	payload types.PaymentPayload,
	requirements types.PaymentRequirements,
) (*x402.SettleResponse, error) {
	verifyResp, err := f.Verify(ctx, payload, requirements)
	if err != nil {
		return &x402.SettleResponse{}, err
	}
	if !verifyResp.IsValid {
		return &x402.SettleResponse{
			Success:     false,
			ErrorReason: verifyResp.InvalidReason,
		}, nil
	}

	hypercorePayload, _ := parsePayload(payload.Payload)
	apiURL := f.getAPIURL(string(requirements.Network))

	payer, err := f.recoverPayer(hypercorePayload.Action, hypercorePayload.Signature)
	if err != nil {
		return &x402.SettleResponse{}, fmt.Errorf("failed to recover payer: %w", err)
	}

	startTime := time.Now()

	submitPayload := map[string]interface{}{
		"action":       hypercorePayload.Action,
		"nonce":        hypercorePayload.Nonce,
		"signature":    hypercorePayload.Signature,
		"vaultAddress": nil,
	}

	body, err := json.Marshal(submitPayload)
	if err != nil {
		return &x402.SettleResponse{}, fmt.Errorf("failed to marshal payload: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, "POST", apiURL+"/exchange", bytes.NewReader(body))
	if err != nil {
		return &x402.SettleResponse{}, fmt.Errorf("failed to create request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return &x402.SettleResponse{}, fmt.Errorf("failed to submit to hyperliquid: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return &x402.SettleResponse{}, fmt.Errorf("hyperliquid API error: %d", resp.StatusCode)
	}

	var apiResp hypercore.HyperliquidAPIResponse
	if err := json.NewDecoder(resp.Body).Decode(&apiResp); err != nil {
		return &x402.SettleResponse{}, fmt.Errorf("failed to decode response: %w", err)
	}

	if apiResp.Status != "ok" {
		return &x402.SettleResponse{
			Success:     false,
			ErrorReason: ErrSettlementFailed,
		}, nil
	}

	txHash, err := f.getTransactionHash(
		ctx,
		apiURL,
		payer,
		hypercorePayload.Action.Destination,
		hypercorePayload.Nonce,
		startTime,
	)
	if err != nil {
		return &x402.SettleResponse{}, fmt.Errorf("failed to get transaction hash: %w", err)
	}

	return &x402.SettleResponse{
		Success:     true,
		Transaction: txHash,
		Network:     x402.Network(requirements.Network),
		Payer:       payer,
	}, nil
}

func (f *ExactHypercoreScheme) recoverPayer(
	action hypercore.HypercoreSendAssetAction,
	signature hypercore.HypercoreSignature,
) (string, error) {
	domain := evm.TypedDataDomain{
		Name:              "HyperliquidSignTransaction",
		Version:           "1",
		ChainID:           big.NewInt(999),
		VerifyingContract: "0x0000000000000000000000000000000000000000",
	}

	typedDataTypes := map[string][]evm.TypedDataField{
		"HyperliquidTransaction:SendAsset": {
			{Name: "hyperliquidChain", Type: "string"},
			{Name: "destination", Type: "string"},
			{Name: "sourceDex", Type: "string"},
			{Name: "destinationDex", Type: "string"},
			{Name: "token", Type: "string"},
			{Name: "amount", Type: "string"},
			{Name: "fromSubAccount", Type: "string"},
			{Name: "nonce", Type: "uint64"},
		},
	}

	// Note: uint64 types in EIP-712 need to be provided as strings
	message := map[string]interface{}{
		"hyperliquidChain": action.HyperliquidChain,
		"destination":      action.Destination,
		"sourceDex":        action.SourceDex,
		"destinationDex":   action.DestinationDex,
		"token":            action.Token,
		"amount":           action.Amount,
		"fromSubAccount":   action.FromSubAccount,
		"nonce":            fmt.Sprintf("%d", action.Nonce),
	}

	hash, err := evm.HashTypedData(domain, typedDataTypes, "HyperliquidTransaction:SendAsset", message)
	if err != nil {
		return "", fmt.Errorf("failed to hash typed data: %w", err)
	}

	rBytes, err := hex.DecodeString(strings.TrimPrefix(signature.R, "0x"))
	if err != nil {
		return "", fmt.Errorf("invalid r value: %w", err)
	}
	sBytes, err := hex.DecodeString(strings.TrimPrefix(signature.S, "0x"))
	if err != nil {
		return "", fmt.Errorf("invalid s value: %w", err)
	}

	v := byte(signature.V)
	if v >= 27 {
		v -= 27
	}

	sig := append(append(rBytes, sBytes...), v)

	pubKey, err := crypto.SigToPub(hash, sig)
	if err != nil {
		return "", fmt.Errorf("failed to recover public key: %w", err)
	}

	address := crypto.PubkeyToAddress(*pubKey)

	return address.Hex(), nil
}

func (f *ExactHypercoreScheme) getTransactionHash(
	ctx context.Context,
	apiURL string,
	user string,
	destination string,
	nonce int64,
	startTime time.Time,
) (string, error) {
	for attempt := 0; attempt < hypercore.TxHashMaxRetries; attempt++ {
		if attempt > 0 {
			time.Sleep(hypercore.TxHashRetryDelay)
		}

		queryPayload := map[string]interface{}{
			"type":      "userNonFundingLedgerUpdates",
			"user":      user,
			"startTime": startTime.Add(-hypercore.TxHashLookbackWindow).UnixMilli(),
		}

		body, err := json.Marshal(queryPayload)
		if err != nil {
			continue
		}

		req, err := http.NewRequestWithContext(ctx, "POST", apiURL+"/info", bytes.NewReader(body))
		if err != nil {
			continue
		}
		req.Header.Set("Content-Type", "application/json")

		resp, err := http.DefaultClient.Do(req)
		if err != nil {
			continue
		}

		respBody, err := io.ReadAll(resp.Body)
		resp.Body.Close()
		if err != nil {
			continue
		}

		var updates []hypercore.LedgerUpdate
		if err := json.Unmarshal(respBody, &updates); err != nil {
			continue
		}

		for _, update := range updates {
			if update.Delta.Type == "send" &&
				update.Delta.Destination != nil &&
				strings.EqualFold(*update.Delta.Destination, destination) &&
				update.Delta.Nonce != nil &&
				*update.Delta.Nonce == nonce {
				return update.Hash, nil
			}
		}
	}

	return "", fmt.Errorf("transaction hash not found after %d attempts", hypercore.TxHashMaxRetries)
}

func parsePayload(payload interface{}) (*hypercore.HypercorePaymentPayload, error) {
	jsonBytes, err := json.Marshal(payload)
	if err != nil {
		return nil, err
	}

	var hypercorePayload hypercore.HypercorePaymentPayload
	if err := json.Unmarshal(jsonBytes, &hypercorePayload); err != nil {
		return nil, err
	}

	return &hypercorePayload, nil
}
